### PR TITLE
Using iterator with fori_loop documentation

### PR DESCRIPTION
**Pull request by IgorWilbert**
For #3567. Please let me know if I made a mistake while editing the documentation.

### CODE REVIEW

File: docs/notebooks/Common_Gotchas_in_JAX.ipynb [hunk 1]
```diff
@@ -1,2631 +1,2676 @@
...
...VRcR11VYIpp6urEKpmTWM1FYZNBI0YqGwBiP1k8gokUxhtUoIrj/epcBSKs08\nuNKksdNAUcOqMzWsgsaIlApBmSSwWUbbQItC6i60eStIgEETg8KUEhhc+kkRM8YYSRo8/ugbub7p\nojuosjWfhd188xh8vjgf97xmdkPiQzgwpch1y3y8zpyWjRiE7w4HwCEs//3TZ87rhrTG99OOTfyF\n8l+XK+f1St4a99NIkcIwRr4sV87bRsmNg4w8DgMmDkm5lJ5PVhqHmGjmhWwuhRgTOW9df7mx08hW\nCtNu5DIvFBotN/bqs8b7KXFpRpFKzsY+RlIIDA2W4JE2S3CYSQyRECK1Nq6tslpDAtylidoqIsaq\nxiaKFYEQGVBMK4XCMM1MoRDECGos654ar4TYSJoxgSl4Xs6ljL7ITAv7sHDU4sSy5oue2nW4gzi1\nqierE0R57NZjcJ7tGy18scjaY8xHNY628dliX1rBgYVM/D/Ze7MmObIkS+9TvfeambvHAiQya5le\npmWGQuH//yMcinARkt0sGVZVduUCIBZfzO6iyge1QBYf5oEzPSNNQdoLBAHAI+ARpqZX9ZzvMEnn\nLm8xSrCF1TNFB3fcqF6oOEkg6eDWFtyFRgoqV6qsXhik6G73r9eSf+lktTuG7iMUQCWswezS4PA4\nMFQi7meE4iygPvHzV2WP2YnfMllk0nWcIsJgBOzcZY+Rj4dSdWNOQqZ8yeVRVxqd1WJmn0vosfvQ\nUCtI7DayR3BmH6DqTBJwG9uM5rEsnXOMHPqoIRtDWJJy0GDxVm2I5fi3krEStIl1dO5KYbbQJFcF\n2Tbul4WHOYrv18LSha+86FaLaJyi4Tb6x0+f+NPrC5jxXFd+e7rjNC9cbiv/9PSRQuLT5caHu5n7\naaGbB093DC63zqyJu/sTAny+nHndNrw71Qe/Ox3pw+hF+efLGe9G7c6Heca1YNJ52kJtubVG2efN\nLkIzxbNQB/hoJHvzzxeGD3pyzrctOkJXTtOM9YpNiefaKCnj3fkuZVQzKSmX3hALQ8idFlJaAOO6\nw1VQOJlCVjbimNpxhjhzUkQnmkelGd6Z5kFtBe8lUgqmimrjbrmypIojJDVetxM1vZLEKSlcbLNA\nJlFdMJzinVmjg48rNKB5txtkBFUheXR5QwbvZeDe96MzH
```
Comment (IgorWilbert) [inline docs/notebooks/Common_Gotchas_in_JAX.ipynb:1597]: ... and ends in this line.

**Comment by gnecula** [comment]
Thank you for helping improve the documentation. My feeling is that the point of not using an iterator deserves to be higher-up in the notebook, where we talk about pure functions. This is because iterators are bad not just in fori_loops, but in pretty much any JAX code that you want to jit, or in any control-flow primitive. Hence, I suggest you move this contents further up, and perhaps make the example independent of fori_loop. You can mention that this (and the rest of purity constraints) apply inside control-flow primitives.

**Comment by IgorWilbert** [comment]
@gnecula thank you for the suggestions! I have updated the description and examples. Please let me know if additional changes are required.

**Review by gnecula — APPROVED** [review]
Thank you!

**Pull request merged**